feat(cli): add option to make custom icon sizes, closes #5121

khuongduy354 opened the pull request:
<!--
Update "[ ]" to "[x]" to check a box

Please make sure to read the Pull Request Guidelines: https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

### What kind of change does this PR introduce?
<!-- Check at least one. If you are introducing a new binding, you must reference an issue where this binding has been proposed, discussed and approved by the maintainers. -->

- [ ] Bugfix
- [X] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?
<!-- If yes, please describe the impact and migration path for existing applications in an attached issue. -->

- [ ] Yes, and the changes were approved in issue #___
- [X] No

### Checklist
- [X] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [ ] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [X] I have added a convincing reason for adding this feature, if necessary

### Other information

khuongduy354 committed:
feat(cli): add option to make custom icon sizes.

khuongduy354 commented:
Feature requested in [this issue](https://github.com/tauri-apps/tauri/issues/5121). Some keynotes:
- Only apply to .ico and .png
- Added --config to find json config file. 
- If no --config provided, use default hardcoded values, as in previous version

khuongduy354 committed:
fix(cli): clippy check.

FabianLars commented:
I don't think there is any value in making it possible to generate custom .ico files. Just like the .icns file it's basically "complete" and there shouldn't be an actual use case requiring different layer sizes.

It's a good start btw! :)

i'll wait with the code nitpicking until it's more feature complete, but one thing already, you can use serde_json::from_reader instead of from_slice. It's a little cleaner and probablyyy faster.

khuongduy354 committed:
refractor(cli): read icon config

FabianLars commented:
Hello 👋 
Sorry for the super long delay, it thought you weren't done and then forgot about it after some time :/

Do you still have interest in working on it or should we take it from here?

A few open points i see at first glance:
- It still supports ICO files. As said earlier i don't think this offers much value because the default ico is already "complete"
- There's still a todo for the file name, but that's imo not important because of my next point
- I don't really like the json approach anymore tbh. I think we should go with a more direct cli interaction. What i have in mind right now would be a cli argument that takes an array of sizes, for example `tauri icon --extra 123,420,1337`, which just reuses the existing png loop, including the default naming scheme. I think letting the users rename the files if needed is not too much to ask for.

khuongduy354 commented:
I was kinda tight on schedule at moment, but i think i can work on it this
weekend, you can leave it to me.

Vào 22:04, T.5, 8 Th12, 2022 Fabian-Lars ***@***.***> đã viết:

> Hello 👋
> Sorry for the super long delay, it thought you weren't done and then
> forgot about it after some time :/
>
> Do you still have interest in working on it or should we take it from here?
>
> A few open points i see at first glance:
>
>    - It still supports ICO files. As said earlier i don't think this
>    offers much value because the default ico is already "complete"
>    - There's still a todo for the file name, but that's imo not important
>    because of my next point
>    - I don't really like the json approach anymore tbh. I think we should
>    go with a more direct cli interaction. What i have in mind right now would
>    be a cli argument that takes an array of sizes, for example tauri icon
>    --extra 123,420,1337, which just reuses the existing png loop,
>    including the default naming scheme. I think letting the users rename the
>    files if needed is not too much to ask for.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tauri-apps/tauri/pull/5246#issuecomment-1342868838>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKBJSIICA7VXTKSALKQCZX3WMH2GNANCNFSM6AAAAAAQRDI434>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

khuongduy354 committed:
feat: remove config json as an option

khuongduy354 committed:
feature: extra flag for custom png sizes

cargo tauri icon icon.png --extra 13,15,17\nthis generate png icons: 13x13.png, 15x15.png,17x17.png

khuongduy354 commented:
I made some changes, can you have a look? @FabianLars . It remove json config, and --extra flags generate custom sizes .png with same naming scheme. 
Should the --extra generate **additional** besides default png sizes, or it should ignore default ones?

FabianLars commented:
> Should the --extra generate additional besides default png sizes, or it should ignore default ones?

In my opinion it should create _additional_ files, and keep building the default at the same time.

khuongduy354 committed:
refractor:--extra make additional icons instead of replace defaults

khuongduy354 commented:
How is it? If you decide to take it, I'll cleanup the comments and test icon files.
@FabianLars

FabianLars commented:
Yeah i like that. Another thing we may wanna look into is de-duplication so that we don't end up generating the same icon twice - that can be added after the cleanup ofc.

khuongduy354 committed:
cleanup: rm comments,unused files

khuongduy354 committed:
fix: deduplicate extra icons

FabianLars commented on tooling/cli/src/icon.rs:
```diff
@@ -41,15 +41,20 @@ pub struct Options {
   let out_dir = options.output.unwrap_or_else(|| tauri_dir().join("icons"));
+  let extra_icon_sizes = options
+    .extra
+    .unwrap_or_else(|| vec![32, 16, 24, 48, 64, 256]);
```
```suggestion
    .unwrap_or_default();
```
sending an empty array is fine. at this point it may be worth to just send the option and handle this in `fn png` instead so that one can get rid of a potentially unused vec.

FabianLars commented on tooling/cli/src/icon.rs:
```diff
@@ -124,7 +129,10 @@ fn ico(source: &DynamicImage, out_dir: &Path) -> Result<()> {
   let mut frames = Vec::new();
 
-  for size in [32, 16, 24, 48, 64, 256] {
+  let sizes = vec![32, 16, 24, 48, 64, 256];
+  let icon_name = "icon.ico".to_string();
+
+  for size in sizes {
```
```suggestion
  for size in [32, 16, 24, 48, 64, 256] {
```

FabianLars commented on tooling/cli/src/icon.rs:
```diff
@@ -144,7 +152,7 @@ fn ico(source: &DynamicImage, out_dir: &Path) -> Result<()> {
   }
 
-  let mut out_file = BufWriter::new(File::create(out_dir.join("icon.ico"))?);
+  let mut out_file = BufWriter::new(File::create(out_dir.join(icon_name))?);
```
```suggestion
  let mut out_file = BufWriter::new(File::create(out_dir.join("icon.ico"))?);
```

FabianLars commented on tooling/cli/src/helpers/config.json:
```diff
@@ -0,0 +1,10 @@
+    "name": "icon.png",
+    "sizes": [32, 128, 256, 512]
+  }
+}
```
file is not needed anymore

FabianLars commented on tooling/cli/src/icon.rs:
```diff
@@ -154,8 +162,16 @@ fn ico(source: &DynamicImage, out_dir: &Path) -> Result<()> {
+    .into_iter()
+    .filter(|x| !sizes.contains(x))
+    .collect::<Vec<u32>>();
+  sizes.append(extra_icon_sizes.as_mut());
```
it may be nicer to do something like this
```rs
  sizes.append(extra_icon_sizes.as_mut());
  sizes.sort_unstable();
  sizes.dedup();
```
what do you think?

khuongduy354 replied:
- I think .filter() takes linear time. Idk how much it cost to cast from vec to iterator and back. 
- Sorting may take longer, and extra cost on dedup()

FabianLars replied:
i don't think performance on a 5 entry vec matters, it's only about the esthetics :D

really doesn't matter, it was just a random thought.

khuongduy354 replied:
Users may input tons of extra sizes, but that may not happen :D . It's just my habit for best practice.

FabianLars commented on tooling/cli/src/icon.rs:
```diff
@@ -154,8 +162,16 @@ fn ico(source: &DynamicImage, out_dir: &Path) -> Result<()> {
-  for size in [32, 128, 256, 512] {
+fn png(source: &DynamicImage, out_dir: &Path, mut extra_icon_sizes: Vec<u32>) -> Result<()> {
+  //remove duplicate extra icons
+  let mut sizes = vec![32, 16, 24, 48, 64, 256];
```
```suggestion
  let mut sizes = vec![32, 128, 256, 512];
```

FabianLars reviewed (changes requested):
Thanks so far! Only some small-ish things left. On top of what's mentioned in the comments we also need a changefile, example: https://github.com/tauri-apps/tauri/blob/dev/.changes/is-minimized.md

khuongduy354 committed:
cleanup(icon):hard-code some values, smallish things.

khuongduy354 committed:
chore: add changefile

FabianLars reviewed (dismissed):
imo it's good as-is, thanks!

2 potential changes/improvements we may wanna think about anyway:
- make the extra_icon_sizes arg an `Option` so that all the new code is basically skipped for most users
- If we want to make users able to generate 256 and 512 icons with the default name (_on top_ of the 2 hard coded names), we'd have to hard code the creation of the 128x128@2x.png and icon.png icons outside the for loop.
At this point we could get rid of the whole dedup logic by also hard coding the removal of 32px and 128px icons from the extra icons vec 🤷

khuongduy354 committed:
feat(icon):seperate default and extra icons,rm dedup

khuongduy354 commented:
I don't really get making extra_icon_sizes as `Option`.

khuongduy354 committed:
Merge branch 'dev' into feat

khuongduy354 commented:
I altered the icon sizes, as you said, it removes the dedup logic, looks better to me. Do u want it or switch back to the old commit ?

FabianLars commented:
> I don't really get making extra_icon_sizes as Option.

Doesn't really matter anymore. Without the dedup logic it would add more visual clutter than its worth.

One last thing: The 32 and 128 icon sizes are missing now.

khuongduy354 commented:
So it's done after switching back to prev commit ?

FabianLars commented:
That would introduce the 256 / 512 naming problem from earlier. Honestly just do something like this again
```rs
for size in [32, 128].into_iter().chain(extra_icon_sizes) {}
```
and call it a day. We had so many iterations that i'm not sure anymore what i'd like to see at this point 😅 

Thanks for following through btw!

khuongduy354 committed:
fix(icon):simplify, default only use 32,128 sizes

khuongduy354 committed:
Merge branch 'feat' of https://github.com/khuongduy354/tauri into feat

FabianLars committed:
re-add default png icons

FabianLars commented:
Sorry that it was such a pain. I'm quite new to this PR reviewing stuff x)

Thank you again!

FabianLars committed:
fix changefile

khuongduy354 commented:
Thx a bunch, Im a newbie and was busy recently, so i cant focus entirely on this, it such a pain. Anyways, thanks a lot for your guide

FabianLars committed:
fix clap derive

lucasfernog committed:
rename flag

lucasfernog committed:
lint

lucasfernog commented:
LGTM, but maybe we should change it to only run the user-specified icon sizes? Like instead of running `tauri icon --extra-png 24 48` to generate all Tauri icons + the extra ones, we could let the user run `tauri icon --png 24 48 --output ./assets/img`.

FabianLars commented:
That was indeed the initial idea from the others, but i thought it feels a bit weird (idk why) so i asked to start with this approach first and wait for other opinions

lucasfernog committed:
only generate the custom sizes

lucasfernog commented:
I've pushed that. We have time to get feedback and revert it if it feels weird :) we still need to audit the next release anyway.
